test/parallel_api/iterator/input_data_sweep.pass.cpp

```diff
@@ -0,0 +1,348 @@
+                                                    std::string("sycl_iterator<") + type_text + std::string(">"));
+        }
+
+        { // usm_shared
+            auto usm_shared = sycl::malloc_shared<T>(n, policy4.queue());
+            //test all modes / wrappers
+            wrap_recurse<__recurse, 0>(policy4, usm_shared, usm_shared + n, counting, copy_out, usm_shared, copy_out,
+                                    counting, trash, std::string("usm_shared<") + type_text + std::string(">"));
+            sycl::free(usm_shared, policy4.queue());
+        }
+
+        { // usm_device
+            auto usm_device = sycl::malloc_device<T>(n, policy5.queue());
+            //test all modes / wrappers
+            wrap_recurse<__recurse, 0>(policy5, usm_device, usm_device + n, counting, copy_out, usm_device, copy_out,
+                                    counting, trash, std::string("usm_device<") + type_text + std::string(">"));
+
+            sycl::free(usm_device, policy5.queue());
+        }
+        sycl::free(copy_out, policy.queue());
+    }
+    else
+    {
+        TestUtils::unsupported_types_notifier(policy.queue().get_device());
+    }
+}
+
+#endif //TEST_DPCPP_BACKEND_PRESENT
+
+int
+main()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
+
+    constexpr size_t n = 10;
+
+    auto q = TestUtils::get_test_queue();
+
+    auto policy = TestUtils::create_test_policy<class Kernel1>(q);
+
+    auto policy1 = TestUtils::create_new_policy_idx<decltype(policy), 0>(policy);
+    auto policy2 = TestUtils::create_new_policy_idx<decltype(policy), 1>(policy);
+    auto policy3 = TestUtils::create_new_policy_idx<decltype(policy), 2>(policy);
+    auto policy4 = TestUtils::create_new_policy_idx<decltype(policy), 3>(policy);
+    auto policy5 = TestUtils::create_new_policy_idx<decltype(policy), 4>(policy);
+    auto policy6 = TestUtils::create_new_policy_idx<decltype(policy), 5>(policy);
+
+    // baseline with no wrapping
+    test<float, 0>(policy1, -666.0f, n, "float");
+    test<double, 0>(policy2, -666.0, n, "double");
+    test<std::uint64_t, 0>(policy3, 999, n, "uint64_t");
+
+    // big recursion step: 1 and 2 layers of wrapping
+    test<std::int32_t, 2>(policy4, -666, n, "int32_t");
+
+    // special cases
+
+    // discard iterator
+    oneapi::dpl::counting_iterator<int> counting(0);
+    oneapi::dpl::discard_iterator discard{};
+    wrap_recurse<1, 0, /*__read =*/false, /*__reset_read=*/false, /*__write=*/true,
+                 /*__check_write=*/false, /*__usable_as_perm_map=*/false, /*__usable_as_perm_src=*/true,
+                 /*__is_reversible=*/true>(policy5, discard, discard + n,
+                                           counting, discard, discard, discard, discard, -666, "discard_iterator");
+
+    // recurse once on perm(perm(usm_shared<int>,count), count)
+    auto copy_out = sycl::malloc_shared<int>(n, policy6.queue());
```

Comment:
You can try to use `usm_data_transfer` class - it's already has `alloc` and `free` calls

Reply:
I've switched to using that class for the data allocation / freeing.  Thanks.

Reply:
ok